controllers/util.go

```diff
@@ -0,0 +1,107 @@
+package controllers
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	. "github.com/vshn/stardog-userrole-operator/api/v1alpha1"
+)
+
+func getCredentials(ctx context.Context, client client.Client, credentials StardogUserCredentialsSpec) (username, password string, err error) {
+	secret := &v1.Secret{}
+	err = client.Get(ctx, types.NamespacedName{Namespace: credentials.Namespace, Name: credentials.SecretRef}, secret)
+	if err != nil {
+		return "", "", fmt.Errorf("cannot retrieve credentials from Secret %s/%s: %v", credentials.Namespace, credentials.SecretRef, err)
+	}
+	user := secret.StringData["username"]
+	if username == "" {
+		return "", "", fmt.Errorf(".data.username in the Secret %s/%s is required", credentials.Namespace, credentials.SecretRef)
+	}
+	pass := secret.StringData["password"]
+	if password == "" {
+		return "", "", fmt.Errorf(".data.password in the Secret %s/%s is required", credentials.Namespace, credentials.SecretRef)
+	}
```

Comment:
You should get the secrets from the base64 encoded `Data`, not the plaintext ones in `StringData`. Best is probably to check both, while `Data` takes precedence.

also double check if an empty string decoded to base64 is also empty or kind of contains characters that represent the decoded "empty" string. I'm not quite sure here

Reply:
I will update what you suggested but if it's empty there's no reason to decode it, as there should be a password and a user